### ISSUE TITLE
Having trouble running app with Netty

### ISSUE THREAD

**Issue by danielfernandez** [bclozel/spring-boot-web-reactive#8]
I'm trying to *bootify* the Thymeleaf + Spring Reactive sandbox application at https://github.com/thymeleaf/thymeleafsandbox-springreactive by means of this starter, and I'm finding trouble trying to make it work with Netty.

I'll try to carefully explain the configuration steps, maybe I'm missing something:

As a first approach, I configured the application based on what I could see at the *sample* application at this same repository (but switching to netty instead of tomcat):

```xml
  <properties>
    ...
    <spring.version>5.0.0.BUILD-SNAPSHOT</spring.version>
    <reactor.version>3.0.0.BUILD-SNAPSHOT</reactor.version>
    <reactor-netty.version>0.5.0.BUILD-SNAPSHOT</reactor-netty.version>
  </properties>

  <dependencies>
  
    <dependency>
      <groupId>org.springframework.boot.experimental</groupId>
      <artifactId>spring-boot-starter-web-reactive</artifactId>
      <exclusions>
        <!-- Excluded because we will be using netty -->
        <exclusion>
          <groupId>org.springframework.boot</groupId>
          <artifactId>spring-boot-starter-tomcat</artifactId>
        </exclusion>
      </exclusions>
    </dependency>

    <dependency>
      <groupId>io.projectreactor</groupId>
      <artifactId>reactor-netty</artifactId>
      <version>${reactor-netty.version}</version>
    </dependency>

    ...

  </dependencies>
```

But when trying to execute this, I got:

```
java.lang.NoSuchMethodError: io.netty.handler.codec.http.HttpRequest.protocolVersion()Lio/netty/handler/codec/http/HttpVersion;
	at reactor.io.netty.http.NettyHttpChannel.<init>(NettyHttpChannel.java:88) ~[reactor-netty-0.5.0.BUILD-SNAPSHOT.jar:na]
	at reactor.io.netty.http.HttpServerChannel.<init>(HttpServerChannel.java:39) ~[reactor-netty-0.5.0.BUILD-SNAPSHOT.jar:na]
	at reactor.io.netty.http.HttpServer.createChannelBridge(HttpServer.java:480) ~[reactor-netty-0.5.0.BUILD-SNAPSHOT.jar:na]
	at reactor.io.netty.http.HttpServer.createChannelBridge(HttpServer.java:60) ~[reactor-netty-0.5.0.BUILD-SNAPSHOT.jar:na]
	at reactor.io.netty.http.NettyHttpServerHandler.channelRead(NettyHttpServerHandler.java:80) ~[reactor-netty-0.5.0.BUILD-SNAPSHOT.jar:na]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-all-4.0.38.Final.jar:4.0.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:334) [netty-all-4.0.38.Final.jar:4.0.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:327) [netty-all-4.0.38.Final.jar:4.0.38.Final]
    ...
```

So I fixed the version of netty to `4.1.3.Final` and added `netty-all` as a dependency (if not, some parts of `reactor-netty` would end up including jars from Netty 4.0.38):

```xml
  <properties>
    ...
    <netty.version>4.1.3.Final</netty.version>
  </properties>

  <dependencies>

    ...

    <dependency>
      <groupId>io.netty</groupId>
      <artifactId>netty-all</artifactId>
      <version>${netty.version}</version>
    </dependency>

    ...

  </dependencies>
```

...but this was to no avail, because this time I got:

```
java.lang.NoClassDefFoundError: reactor/ipc/netty/http/HttpChannel
	at org.springframework.http.server.reactive.ReactorHttpHandlerAdapter.apply(ReactorHttpHandlerAdapter.java:33) ~[spring-web-5.0.0.BUILD-SNAPSHOT.jar:5.0.0.BUILD-SNAPSHOT]
	at reactor.io.netty.http.HttpServer.lambda$doStart$48(HttpServer.java:392) ~[reactor-netty-0.5.0.BUILD-SNAPSHOT.jar:na]
	at reactor.io.netty.http.NettyHttpServerHandler.channelRead(NettyHttpServerHandler.java:89) ~[reactor-netty-0.5.0.BUILD-SNAPSHOT.jar:na]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:372) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:358) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:350) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:435) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:250) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:372) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:358) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:350) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:372) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:358) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:123) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:571) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:512) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:426) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:398) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:877) [netty-all-4.1.3.Final.jar:4.1.3.Final]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_66]
Caused by: java.lang.ClassNotFoundException: reactor.ipc.netty.http.HttpChannel
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381) ~[na:1.8.0_66]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[na:1.8.0_66]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[na:1.8.0_66]
	... 24 common frames omitted
```

Given it was complaining about an `.ipc.` class, I tried to use the `reactor-netty` dependency in the `io.projectreactor.ipc` groupID:

```xml
  <dependencies>

    ...

    <dependency>
      <groupId>io.projectreactor.ipc</groupId>
      <artifactId>reactor-netty</artifactId>
      <version>${reactor-netty.version}</version>
    </dependency>

    ...

  </dependencies>
```

...but this time the problem was Netty's HttpServer was not being *autoconfigured* at all:

```
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.springframework.boot.maven.AbstractRunMojo$LaunchRunner.run(AbstractRunMojo.java:483)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.springframework.context.ApplicationContextException: Unable to start embedded container; nested exception is org.springframework.context.ApplicationContextException: Unable to start ReactiveWebApplicationContext due to missing ReactiveHttpServerFactory bean.
	at org.springframework.boot.context.embedded.ReactiveWebApplicationContext.onRefresh(ReactiveWebApplicationContext.java:52)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:535)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:369)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:313)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1185)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1174)
	at thymeleafsandbox.springreactive.application.Application.main(Application.java:33)
	... 6 more
Caused by: org.springframework.context.ApplicationContextException: Unable to start ReactiveWebApplicationContext due to missing ReactiveHttpServerFactory bean.
	at org.springframework.boot.context.embedded.ReactiveWebApplicationContext.getReactiveHttpServerFactory(ReactiveWebApplicationContext.java:84)
	at org.springframework.boot.context.embedded.ReactiveWebApplicationContext.createReactiveHttpServer(ReactiveWebApplicationContext.java:70)
	at org.springframework.boot.context.embedded.ReactiveWebApplicationContext.onRefresh(ReactiveWebApplicationContext.java:49)
	... 13 more
```

**What can be wrong here?**

By the way, I also tried using version `2.5.0.BUILD-SNAPSHOT` of `reactor-netty`, which was what I was previously using (*why `0.5.0.BUILD-SNAPSHOT` is newer in the Spring milestone repo than `2.5.0.BUILD-SNAPSHOT`?*):

```xml
  <properties>
    ...
    <reactor-netty.version>2.5.0.BUILD-SNAPSHOT</reactor-netty.version>
    ...
  </properties>
```

Again, no *autoconfiguration* in this case:

```
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.springframework.boot.maven.AbstractRunMojo$LaunchRunner.run(AbstractRunMojo.java:483)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.springframework.context.ApplicationContextException: Unable to start embedded container; nested exception is org.springframework.context.ApplicationContextException: Unable to start ReactiveWebApplicationContext due to missing ReactiveHttpServerFactory bean.
	at org.springframework.boot.context.embedded.ReactiveWebApplicationContext.onRefresh(ReactiveWebApplicationContext.java:52)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:535)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:369)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:313)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1185)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1174)
	at thymeleafsandbox.springreactive.application.Application.main(Application.java:33)
	... 6 more
Caused by: org.springframework.context.ApplicationContextException: Unable to start ReactiveWebApplicationContext due to missing ReactiveHttpServerFactory bean.
	at org.springframework.boot.context.embedded.ReactiveWebApplicationContext.getReactiveHttpServerFactory(ReactiveWebApplicationContext.java:84)
	at org.springframework.boot.context.embedded.ReactiveWebApplicationContext.createReactiveHttpServer(ReactiveWebApplicationContext.java:70)
	at org.springframework.boot.context.embedded.ReactiveWebApplicationContext.onRefresh(ReactiveWebApplicationContext.java:49)
	... 13 more
```

Thanks!

**Comment by danielfernandez** [comment]
Apparently fixed now thanks to https://github.com/bclozel/spring-boot-web-reactive/commit/78d6aa6cf6d2b706787169d3d1486fb3f566ff96. Also, great to have a BOM and a parent pom now, thanks @bclozel @snicoll !

This can be closed now (unless you have any organisational reasons to keep it open…)

**Comment by bclozel** [comment]
Ha, I was about to respond that.
You were basically running into all the possible dependency issues.
You need to update your build file to import that BOM and remove versions managed by it.

Something like this:

```
<dependencyManagement>
	<dependencies>
		<dependency>
			<groupId>org.springframework.boot.experimental</groupId>
			<artifactId>spring-boot-dependencies-web-reactive</artifactId>
			<version>0.1.0.BUILD-SNAPSHOT</version>
			<type>pom</type>
			<scope>import</scope>
		</dependency>
	</dependencies>
</dependencyManagement>
```

I'll add more documentation to the README now that things settled.

**Comment by danielfernandez** [comment]
> You need to update your build file to import that BOM and remove versions managed by it.

Do I need that even once I extend from the parent pom? https://github.com/thymeleaf/thymeleafsandbox-springreactive/blob/f4e09f3b3fe950dcc3a91b76df926f28f3d9e2fd/pom.xml#L33-L39

**Comment by bclozel** [comment]
That was one of the ways of doing it, but for now we favor the regular Boot starter.
You can take a look at what start-development.cfapps.io is generating (and soon the production instance).

The sample application inside this repository is more a sanity check than a real sample you copy/paste.

**Comment by snicoll** [comment]
@danielfernandez `spring-boot-web-reactive-parent` is not an equivalent of `spring-boot-starter-parent` so I wouldn't use that. Do you think we should provide that (hint: we would have to copy/paste everything from `spring-boot-starter-parent` so I am not super keen of doing that).

**Comment by danielfernandez** [comment]
OK, using the regular Spring Boot starter makes sense to me in order to favour cleaner composition of starters and BOMs. I just did that to my application, applying a setup similar to the apps generated by `start-development.cfapps.io`. Thanks.

One comment, out of curiosity: given the dependencies affected by the `reactor.version` property and also most of the ones affected by `spring.version` are left out of the dependency management section of `spring-boot-dependencies-web-reactive`, I needed to manually set these two properties (Reactor `3.0.0.BUILD-SNAPSHOT` and Spring `5.0.0.BUILD-SNAPSHOT`). That's also what is done at the applications generated by `start-development.cfapps.io`. 

My question is… was this done on purpose because this `spring-boot-dependencies-web-reactive` is not considered to be the adequate *source of truth* for these versions (should be the base Spring Boot BOM instead), and so this will be like that for now until in the future Spring Boot upgrades to Spring 5?

And if the answer is *"yes, this is just because `spring-web-reactive` cannot exist at the base Spring Boot BOM yet because it doesn't exist in 4.3"*, and so this is a sort of *temporary hack*… why are not the `spring-core`, `spring-framework-bom` and `io.projectreactor:*` dependencies managed there too as a part of that temporary hack?

Sorry for the amount of doubts/questions, and the length of them 😊

**Comment by snicoll** [comment]
This is actually a very good question. The only reason those version overrides are there is to [work around a bug in Maven that will be fixed in 3.4](https://issues.apache.org/jira/browse/MNG-5971).

It is the adequate source of truth but we have no choice at the moment I am afraid. You don't need to do this with Gradle (generate a Gradle project and you'll see).

I'll actually submit a PR later today to rename that parent to something else to avoid the confusion.

We'll keep that infrastructure and when Spring 5 will be integrated in Boot 2.0, you'll be able to resume "normal" operations as far as dependency management goes.

**Comment by danielfernandez** [comment]
Understood. Thanks @snicoll !
